cmd/geth/snapshot.go

```diff
@@ -1020,3 +1041,162 @@ func checkAccount(ctx *cli.Context) error {
 	log.Info("Checked the snapshot journalled storage", "time", common.PrettyDuration(time.Since(start)))
 	return nil
 }
+
+func traversalVar(ctx *cli.Context) error {
+	stack, _ := makeConfigNode(ctx)
+	defer stack.Close()
+
+	_, db, root, err := parseDumpConfig(ctx, stack)
+	if err != nil {
+		return err
+	}
+	defer db.Close()
+	triedb := utils.MakeTrieDatabase(ctx, stack, db, false, true, false)
+	defer triedb.Close()
+
+	snapConfig := snapshot.Config{
+		CacheSize:  256,
+		Recovery:   false,
+		NoBuild:    true,
+		AsyncBuild: false,
+	}
+	triesInMemory := ctx.Uint64(utils.TriesInMemoryFlag.Name)
+	snaptree, err := snapshot.New(snapConfig, db, triedb, root, int(triesInMemory), false)
+	snapshot := snaptree.Snapshot(root)
+	if err != nil {
+		return err
+	}
+	contractAddress := common.HexToAddress("0xC806e70a62eaBC56E3Ee0c2669c2FF14452A9B3d")
+	emptyKeyCount := 0
+	varCount := 0
+	arrayCount := 0
+	arrayVarCount := 0
+	for i := 0; ; i++ {
+		key := common.BigToHash(big.NewInt(int64(i)))
+		enc, encErr := snapshot.Storage(crypto.Keccak256Hash(contractAddress[:]), crypto.Keccak256Hash(key.Bytes()))
+		if encErr != nil {
+			log.Info(fmt.Sprintf("key: %s, storage error: %v", key.String(), encErr))
+			continue
+		}
+		if len(enc) > 0 {
+			log.Info(common.BytesToHash(enc).String())
+			emptyKeyCount = 0
+			if isArr, arrLen := isArray(snapshot, contractAddress, key); isArr {
+				arrayVarCount += arrLen
+				arrayCount++
+			} else {
+				varCount++
+			}
+
+		} else {
+			emptyKeyCount++
+		}
+		if emptyKeyCount == 20 {
+			break
+		}
+	}
+	log.Info("array count:", arrayCount, "array item count", arrayVarCount, "var count", varCount)
+	return nil
+}
+
+func isArray(snap snapshot.Snapshot, contractAddress common.Address, slotIdx common.Hash) (result bool, slotLen int) {
+	start := crypto.Keccak256Hash(slotIdx.Bytes())
+	log.Info("start", start.String())
+	emptyValue := 0
+	hasValue := false
+	for i := 0; ; i++ {
+		idx := common.BigToHash(big.NewInt(0).Add(start.Big(), big.NewInt(int64(i))))
+		log.Info("idx:", idx.String(), "big", idx.Big().String())
+		enc, err := snap.Storage(crypto.Keccak256Hash(contractAddress[:]), crypto.Keccak256Hash(idx.Bytes()))
+		if err != nil {
+			log.Info("storage error", err.Error(), "slot", idx.String())
+			continue
```

Comment:
handler the error better - if every `i` get error, the codes will loop forever

Reply:
fixed